File: Visual Studio LIghtswitch 2012/Chapter13/Chapter13/ApressExtensionVB/ApressExtensionVB/ApressExtensionVB.Common/MaxIntegerValidationFactory.vb
Imports System.ComponentModel.Composition
Imports Microsoft.LightSwitch.Runtime.Rules
Imports Microsoft.LightSwitch.Model



<Export(GetType(IValidationCodeFactory))>
<ValidationCodeFactory("ApressExtensionVB:@MaxIntegerValidationId")>
Public Class MaxIntegerValidationFactory
    Implements IValidationCodeFactory

    Public Function Create(
       modelItem As Microsoft.LightSwitch.Model.IStructuralItem,
       attributes As System.Collections.Generic.IEnumerable(
          Of Microsoft.LightSwitch.Model.IAttribute)) As Microsoft.LightSwitch.Runtime.Rules.IAttachedValidation Implements Microsoft.LightSwitch.Runtime.Rules.IValidationCodeFactory.Create

        If Not IsValid(modelItem) Then
            Throw New InvalidOperationException("Unsupported data type.")
        End If
        Return New MaxIntegerValidation(attributes)
    End Function

    Public Function IsValid(modelItem As Microsoft.LightSwitch.Model.IStructuralItem) As Boolean Implements Microsoft.LightSwitch.Runtime.Rules.IValidationCodeFactory.IsValid

        Dim nullableType As INullableType = TryCast(modelItem, INullableType)

        ' Get underlying type if it is a INullableType.
        modelItem =
           If(nullableType IsNot Nothing, nullableType.UnderlyingType, modelItem)

        ' Ensure the type is an integer type, or that a type inherits from it.
        While TypeOf modelItem Is ISemanticType
            If String.Equals(DirectCast(modelItem, ISemanticType).Id,
               "ApressExtensionVB:DurationType",
                  StringComparison.Ordinal) Then
                Return True
            End If
            modelItem = DirectCast(modelItem, ISemanticType).UnderlyingType
        End While
        ' Don't apply the validation if the conditions aren't met
        Return False

    End Function
End Class

File: Visual Studio LIghtswitch 2012/Chapter13/Chapter13/ApressExtensionCS/ApressExtensionCS/ApressExtensionCS.Common/MaxIntegerValidationFactory.cs
//You can use and redistribute the code from this book (and sample application) for non-commercial and 
//most commercial purposes as long as you acknowledge the source and authorship. 
//You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
//The acknowledgment should include author, title, publisher, and ISBN. 
//An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

using System;
using System.Collections.Generic;
using System.ComponentModel.Composition;
using System.Linq;
using Microsoft.LightSwitch;
using Microsoft.LightSwitch.Model;
using Microsoft.LightSwitch.Runtime.Rules;

//Listing 13-4. Creating the Validation Factory Code
namespace ApressExtensionCS
{
    [Export(typeof(IValidationCodeFactory))]
    [ValidationCodeFactory("ApressExtensionCS:@MaxIntegerValidationId")]      
    public class MaxIntegerValidationFactory : IValidationCodeFactory
    {
        public IAttachedValidation Create(IStructuralItem modelItem,
           IEnumerable<IAttribute> attributes)
        {
            // Enusre that the data type is a positive integer semantic
            // type (or nullable positive integer)
            if (!IsValid(modelItem))
            {
                throw new InvalidOperationException("Unsupported data type.");
            }
            return new MaxIntegerValidation (attributes);
        }

        public bool IsValid(IStructuralItem modelItem)
        {
            INullableType nullableType = modelItem as INullableType;

            // Get underlying type if it is an INullableType.
            modelItem = 
               null != nullableType ? nullableType.UnderlyingType : modelItem;

            // Ensure the type matches the business type, 
            // or the underlying type
            while (modelItem is ISemanticType)
            {
                if (String.Equals(((ISemanticType)modelItem).Id,
                   "ApressExtensionCS:DurationType",                        
                    StringComparison.Ordinal))
                {
                    return true;
                }
                modelItem = ((ISemanticType)modelItem).UnderlyingType;
            }

            //Don't apply the validation if the conditions aren't met
            return false;
        }
    }
}
